src/asset/asset-list-loader.js

```diff
@@ -0,0 +1,215 @@
+        }
+    }
+
+    /**
+     * Sets a callback which will be called when all assets in the list have been loaded or failed to load.
+     *
+     * @param {Function} done - Callback called when all assets in the list are loaded.
+     * @param {object} [scope] - Scope to use when calling callback.
+     */
+    ready(done, scope = this) {
+        if (this._loaded) {
+            done.call(scope, this._assets);
+        } else {
+            this.once("load", function (assets) {
+                done.call(scope, assets);
+            });
+        }
+    }
+
+    // called when all assets are loaded
+    _loadingComplete() {
+        if (this._loaded) return;
+        this._loaded = true;
+        this._registry.off("load", this._onLoad, this);
+        this._registry.off("error", this._onError, this);
+
+        if (this._failed && this._failed.length) {
+            if (this._callback) {
+                this._callback.call(this._scope, "Failed to load some assets", this._failed);
+            }
+            this.fire("error", this._failed);
+        } else {
+            if (this._callback) {
+                this._callback.call(this._scope);
+            }
+            this.fire("load", this._assets);
+        }
+    }
+
+    // called when an (any) asset is loaded
+    _onLoad(asset) {
+        // check this is an asset we care about
+        if (this._loadingAssets.has(asset)) {
+            this._count++;
+            this.fire("progress", asset);
+        }
+
+        if (this._count === this._loadingAssets.size) {
+            // call next tick because we want
+            // this to be fired after any other
+            // asset load events
+            setTimeout(() => {
+                this._loadingComplete(this._failed);
+            }, 0);
+        }
+    }
+
+    // called when an asset fails to load
+    _onError(err, asset) {
+        // check this is an asset we care about
+        if (this._loadingAssets.has(asset)) {
+            this._count++;
+            this._failed.push(asset);
+        }
+
+        if (this._count === this._loadingAssets.size) {
+            // call next tick because we want
+            // this to be fired after any other
+            // asset load events
+            setTimeout(() => {
+                this._loadingComplete(this._failed);
+            }, 0);
+        }
+    }
+
+    // called when a expected asset is added to the asset registry
+    _onAddAsset(asset) {
+        // remove from waiting list
+        const index = this._waitingAssets.indexOf(asset);
+        if (index >= 0) {
+            this._waitingAssets.splice(index, 1);
+        }
+
+        this._assets.push(asset);
+        const l = this._assets.length;
+        for (let i = 0; i < l; i++) {
+            asset = this._assets[i];
+            if (!asset.loaded) {
+                this._loadingAssets.add(asset);
+                this._registry.load(asset);
+            }
+        }
+    }
```

Comment:
it feels odd this function does not only handle the asset that was added, but loops over all assets - is that required?